File: apps/software-ideas/src/app/services/implementation.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { Implementation } from '../models/implementation.model';
import { Recommendation } from '../models/recommendation.model';
import { Reply } from '../models/reply.model';

@Injectable({
  providedIn: 'root'
})
export class ImplementationService {

  public implementationForEditing: Implementation;

  private _serviceURL = "http://localhost:8080/implementation/";

  constructor(
    private readonly _http: HttpClient
  ) { }

  getImplementations(categoryName?: string): Observable<Implementation[]> {
    let urlParameters = '';
    if (categoryName) {
      urlParameters += `categoryName=${categoryName}`;
    }  
    return this._http.get<Implementation[]>(`${this._serviceURL}getImplementations?${urlParameters}`);
  }

  getImplementation(name: string): Observable<Implementation> {
    return this._http.get<Implementation>(`${this._serviceURL}getImplementation?name=${encodeURIComponent(name)}`);
  }

  postImplementation(implementation: Implementation): Observable<Implementation> {
    return this._http.post<Implementation>(`${this._serviceURL}postImplementation`, implementation);
  }

  postVote(implementationId: number, innovatorId: number, up: boolean): Observable<number> {
    const parameters = `implementationId=${implementationId}&innovatorId=${innovatorId}&up=${up}`;
    return this._http.post<number>(`${this._serviceURL}postVote?${parameters}`, null);
  }

  postRecommendation(recommendation: Recommendation): Observable<Recommendation> {
    return this._http.post<Recommendation>(`${this._serviceURL}postRecommendation`, recommendation);
  }

  postRecommendationVote(recommendationId: number, innovatorId: number, up: boolean): Observable<number> {
    const parameters = `recommendationId=${recommendationId}&innovatorId=${innovatorId}&up=${up}`;
    return this._http.post<number>(`${this._serviceURL}postRecommendationVote?${parameters}`, null);
  }

  postRecommendationReply(reply: Reply): Observable<Reply> {
    return this._http.post<Reply>(`${this._serviceURL}postRecommendationReply`, reply);
  }
}

File: services/users-backlog/src/main/java/users_backlog/services/ImplementationService.java
package users_backlog.services;

import java.util.List;
import java.util.logging.Logger;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import users_backlog.dao.CategoryDAO;
import users_backlog.dao.IdeaDAO;
import users_backlog.dao.ImplementationDAO;
import users_backlog.dao.ProductDAO;
import users_backlog.models.Implementation;
import users_backlog.models.Recommendation;
import users_backlog.models.Reply;

@Service
public class ImplementationService {

    private static final Logger log = Logger.getLogger(ImplementationService.class.getName());

    @Autowired IdeaDAO ideaDAO;
    @Autowired ImplementationDAO implementationDAO;
    @Autowired CategoryDAO categoryDAO;
    @Autowired ProductDAO productDAO;

    public List<Implementation> getImplementations(String categoryName) {
        List<Implementation> implementations = implementationDAO.getImplementations(categoryName);
        for (Implementation implementation: implementations) {
            implementation.setCategories(categoryDAO.getCategories(implementation));
        }
        return implementations;
    }

    public Implementation getImplementation(final long id) {
        Implementation implementation = implementationDAO.getImplementation(id);
        if (implementation != null) {
            implementation.setCategories(categoryDAO.getCategories(implementation));
            implementation.setProducts(productDAO.getProducts(implementation));
            implementation.setIdeas(ideaDAO.getIdeas(implementation));
        }
        return implementation;
    }

    public Implementation getImplementation(final String name) {
        Implementation implementation = implementationDAO.getImplementation(name);
        if (implementation != null) {
            implementation.setCategories(categoryDAO.getCategories(implementation));
            implementation.setProducts(productDAO.getProducts(implementation));
            implementation.setIdeas(ideaDAO.getIdeas(implementation));
        }
        return implementation;
    }

    public Implementation postImplementation(final Implementation implementation) {
        categoryDAO.postCategories(implementation.getCategories());
        implementationDAO.postImplementation(implementation);
        implementation.setCategories(categoryDAO.getCategories(implementation));
        return implementation;
    }

    public Long postVote(Long implementationId, Long innovatorId, Boolean up) {
        return implementationDAO.postVote(implementationId, innovatorId, up);
    }

    public Recommendation postRecommendation(Recommendation recommendation) {
        return implementationDAO.postRecommendation(recommendation);
    }

    public Long postRecommendationVote(Long recommendationId, Long innovatorId, Boolean up) {
        return implementationDAO.postRecommendationVote(recommendationId, innovatorId, up);
    }

    public Reply postRecommendationReply(Reply reply) {
        return implementationDAO.postRecommendationReply(reply);
    }

}